File: tests/golden/expected/07-enums-namespaces.go
package main

import "fmt"

type Direction int

const (
	DirectionUp Direction = iota
	DirectionDown
	DirectionLeft
	DirectionRight
)

func (d Direction) String() string {
	switch d {
	case DirectionUp:
		return "Up"
	case DirectionDown:
		return "Down"
	case DirectionLeft:
		return "Left"
	case DirectionRight:
		return "Right"
	default:
		return "Unknown"
	}
}

type Status string

const (
	StatusPending  Status = "PENDING"
	StatusApproved Status = "APPROVED"
	StatusRejected Status = "REJECTED"
)

type Mixed interface {
	isMixed()
}

type MixedNo struct{}

func (MixedNo) isMixed() {}

const MixedNoValue = 0

type MixedYes struct{}

func (MixedYes) isMixed() {}

const MixedYesValue = "YES"

type LogLevel int

const (
	LogLevelDebug LogLevel = iota
	LogLevelInfo
	LogLevelWarning
	LogLevelError
)

func Move(direction Direction) string {
	switch direction {
	case DirectionUp:
		return "Moving up"
	case DirectionDown:
		return "Moving down"
	case DirectionLeft:
		return "Moving left"
	case DirectionRight:
		return "Moving right"
	default:
		return "Unknown direction"
	}
}

type Utils struct{}

func (Utils) FormatDate(date string) string {
	return date
}

func (Utils) ParseDate(str string) string {
	return str
}

const UtilsVERSION = "1.0.0"

type UtilsConfig struct {
	Debug   bool
	Timeout int
}

type UtilsLogger struct {
	prefix string
}

func NewUtilsLogger(prefix string) *UtilsLogger {
	return &UtilsLogger{prefix: prefix}
}

func (l *UtilsLogger) Log(message string) {
	fmt.Printf("[%s] %s\n", l.prefix, message)
}

type UtilsInternal struct{}

func (UtilsInternal) SecretFunction() string {
	return "internal"
}

var (
	logger  = NewUtilsLogger("App")
	dateStr = Utils{}.FormatDate("")
	config  = UtilsConfig{
		Debug:   true,
		Timeout: 5000,
	}
)

type MergedNamespaceData struct {
	Id    string
	Value float64
}

func MergedNamespaceCreateData(id string, value float64) MergedNamespaceData {
	return MergedNamespaceData{Id: id, Value: value}
}

type MergedNamespaceDataManager struct {
	data []MergedNamespaceData
}

func NewMergedNamespaceDataManager() *MergedNamespaceDataManager {
	return &MergedNamespaceDataManager{
		data: make([]MergedNamespaceData, 0),
	}
}

func (dm *MergedNamespaceDataManager) Add(item MergedNamespaceData) {
	dm.data = append(dm.data, item)
}

func (dm *MergedNamespaceDataManager) GetAll() []MergedNamespaceData {
	result := make([]MergedNamespaceData, len(dm.data))
	copy(result, dm.data)
	return result
}

func GetDirectionName(dir Direction) string {
	return dir.String()
}

func GetDirectionValue(name string) (Direction, bool) {
	switch name {
	case "Up":
		return DirectionUp, true
	case "Down":
		return DirectionDown, true
	case "Left":
		return DirectionLeft, true
	case "Right":
		return DirectionRight, true
	default:
		return 0, false
	}
}

func ProcessStatus(status string) {
	fmt.Printf("Processing status: %s\n", status)
}

type FileAccess int

const (
	FileAccessNone      FileAccess = 0
	FileAccessRead      FileAccess = 1 << 1
	FileAccessWrite     FileAccess = 1 << 2
	FileAccessReadWrite FileAccess = FileAccessRead | FileAccessWrite
	FileAccessAdmin     FileAccess = FileAccessRead | FileAccessWrite | (1 << 3)
)

func HasAccess(current FileAccess, required FileAccess) bool {
	return (current & required) == required
}

File: tests/golden/07-enums-namespaces.ts
/**
 * 測試 07: Enum 與 Namespace
 */

// 數字 enum
enum Direction {
  Up,
  Down,
  Left,
  Right
}

// 字串 enum
enum Status {
  Pending = 'PENDING',
  Approved = 'APPROVED',
  Rejected = 'REJECTED'
}

// 混合 enum
enum Mixed {
  No = 0,
  Yes = 'YES'
}

// const enum
const enum LogLevel {
  Debug = 0,
  Info = 1,
  Warning = 2,
  Error = 3
}

// 使用 enum
function move(direction: Direction): string {
  switch (direction) {
    case Direction.Up:
      return 'Moving up';
    case Direction.Down:
      return 'Moving down';
    case Direction.Left:
      return 'Moving left';
    case Direction.Right:
      return 'Moving right';
    default:
      return 'Unknown direction';
  }
}

// Namespace
namespace Utils {
  export function formatDate(date: Date): string {
    return date.toISOString();
  }

  export function parseDate(str: string): Date {
    return new Date(str);
  }

  export const VERSION = '1.0.0';

  export interface Config {
    debug: boolean;
    timeout: number;
  }

  export class Logger {
    constructor(private prefix: string) {}

    log(message: string): void {
      console.log(`[${this.prefix}] ${message}`);
    }
  }

  // 內部 namespace
  export namespace Internal {
    export function secretFunction(): string {
      return 'internal';
    }
  }
}

// 使用 namespace
const logger = new Utils.Logger('App');
const dateStr = Utils.formatDate(new Date());
const config: Utils.Config = {
  debug: true,
  timeout: 5000
};

// Namespace 合併
namespace MergedNamespace {
  export interface Data {
    id: string;
    value: number;
  }
}

namespace MergedNamespace {
  export function createData(id: string, value: number): Data {
    return { id, value };
  }

  export class DataManager {
    private data: Data[] = [];

    add(item: Data): void {
      this.data.push(item);
    }

    getAll(): Data[] {
      return [...this.data];
    }
  }
}

// Enum 作為型別
type DirectionType = keyof typeof Direction;

function getDirectionName(dir: Direction): DirectionType | undefined {
  return Direction[dir] as DirectionType;
}

// Reverse mapping
function getDirectionValue(name: string): Direction | undefined {
  return (Direction as any)[name];
}

// Enum 與 Union type 的結合
type StatusUnion = `${Status}`;

function processStatus(status: StatusUnion): void {
  console.log(`Processing status: ${status}`);
}

// Computed enum members
enum FileAccess {
  None = 0,
  Read = 1 << 1,
  Write = 1 << 2,
  ReadWrite = Read | Write,
  Admin = Read | Write | 1 << 3
}

function hasAccess(current: FileAccess, required: FileAccess): boolean {
  return (current & required) === required;
}

export {
  Direction,
  Status,
  Mixed,
  LogLevel,
  move,
  Utils,
  MergedNamespace,
  getDirectionName,
  getDirectionValue,
  processStatus,
  FileAccess,
  hasAccess
};